tests/test_video.py

```diff
@@ -0,0 +1,113 @@
+from unittest import mock
+import os.path as osp
+
+import cv2
+import numpy as np
+import pytest
+
+from datumaro.components.media import Video
+from datumaro.util.scope import on_exit_do, scoped
+from datumaro.util.test_utils import TestDir
+
+from .requirements import Requirements, mark_requirement
+
+
+def make_sample_video(path, frames=4, frame_size=(10, 20), fps=25.0):
+    """
+    frame_size is (H, W), only even sides
+    """
+
+    writer = cv2.VideoWriter(path, frameSize=tuple(frame_size[::-1]),
+        fps=float(fps), fourcc=cv2.VideoWriter_fourcc(*'MJPG'))
+
+    for _ in range(frames):
+        # Apparently, only uint8 values are supported, but not floats
+        writer.write(np.ones((*frame_size, 3), dtype=np.uint8) * 255)
+
+    writer.release()
+
+@pytest.fixture(scope='module')
+def fxt_sample_video():
+    with TestDir() as test_dir:
+        video_path = osp.join(test_dir, 'video.avi')
+        make_sample_video(video_path, frame_size=(4, 6))
+
+        yield video_path
+
+class VideoTest:
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_read_video(self, fxt_sample_video):
+        video = Video(fxt_sample_video)
+        on_exit_do(video.close)
+
+        assert 4 == video.frame_count
+        assert (4, 6) == video.frame_size
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_read_frames_sequentially(self, fxt_sample_video):
+        video = Video(fxt_sample_video)
+        on_exit_do(video.close)
+
+        for idx, frame in enumerate(video):
+            assert frame.size == video.frame_size
+            assert frame.index == idx
+            assert id(frame.video) == id(video)
+            assert frame == np.ones((*video.frame_size, 3)) * 255
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_read_frames_randomly(self, fxt_sample_video):
+        video = Video(fxt_sample_video)
+        on_exit_do(video.close)
+
+        for idx in {1, 3, 2, 0, 3}:
+            frame = video[idx]
+            assert frame.index == idx
+            assert frame == np.ones((*video.frame_size, 3)) * 255
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    @scoped
+    def test_can_skip_frames_between(self, fxt_sample_video):
+        video = Video(fxt_sample_video, step=2)
+        on_exit_do(video.close)
+
+        assert 2 == video.frame_count
```

Comment:
With all of these skipping tests, it would also be useful to check that the returned frames are actually the frames that are expected.

Reply:
Done